cdap-app-fabric/src/main/java/io/cdap/cdap/app/deploy/AppDeletionSubscriberService.java

```diff
@@ -92,36 +92,24 @@ protected TransactionRunner getTransactionRunner() {
     return transactionRunner;
   }
 
-  @Nullable
   @Override
-  protected String loadMessageId(StructuredTableContext context) throws Exception {
-    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
-    return appMetadataStore.retrieveSubscriberState(getTopicId().getTopic(), SUBSCRIBER);
-  }
-
-  @Override
-  protected void storeMessageId(StructuredTableContext context, String messageId) throws Exception {
-    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
-    appMetadataStore.persistSubscriberState(getTopicId().getTopic(), SUBSCRIBER, messageId);
-  }
-
-  @Override
-  protected void processMessages(StructuredTableContext structuredTableContext,
-                                 Iterator<ImmutablePair<String, AppDeletionMessage>> messages) throws Exception {
+  protected void processMessages(Iterator<ImmutablePair<String, AppDeletionMessage>> messages) throws Exception {
     while (messages.hasNext()) {
       ImmutablePair<String, AppDeletionMessage> pair = messages.next();
       AppDeletionMessage message = pair.getSecond();
-      ApplicationId appId = message.getPayload(GSON, ApplicationId.class);
-      removeApp(structuredTableContext, appId);
+      ApplicationId appId = message.getApplicationId();
+      removeApp(appId);
     }
   }
 
-  private void removeApp(StructuredTableContext context, ApplicationId id) throws IOException {
-    getAppStateTable(context).deleteAll(id.getNamespaceId(), id.getApplication());
-    AppMetadataStore metaStore = getAppMetadataStore(context);
-    metaStore.deleteApplication(id.getAppReference());
-    metaStore.deleteApplicationEditRecord(id.getAppReference());
-    metaStore.deleteProgramHistory(id.getAppReference());
+  private void removeApp(ApplicationId id) throws IOException {
+    TransactionRunners.run(transactionRunner, context -> {
```

Comment:
@albertshau @tivv does running these methods in a transaction ensure safety

Reply:
the parent class is already starting a transaction and then calling processMessages. Should not be doing this here, it will be a nested transaction.

Reply:
got it